# Part 1. Equation of a Slime

How many late days are you using for this assignment? 1 day and 2 hours

In [1]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
   

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [8]:
# Take the pandas dataset and split it into our features (X) and label (y)
x = data[['Mols KCL', 'Temperature °C']]
y = data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [22]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = np.array([[5, 30]])  # Example values for KCl and Temperature
prediction = model.predict(sample_point)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
# Create the LaTeX formatted equation
coef = model.coef_
intercept = model.intercept_

# Print the coefficients and intercept for clarity
print(coef, intercept)
print(score)


[1032.69506649  866.14641337] -409391.4795834082
0.8552472077276096


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Latex equation = 
$$
\text{Size} = 1032.69507 \cdot \text{KCl} + 866.14641 \cdot \text{Temperature} - 409391.47958
$$



Write the linear equation of a slime: (example equation: $E = mc^2$)

Report on score and explain meaning: The R² score of the linear regression model is R² = 0.85. 
The R² score, also known as the coefficient of determination, indicates the proportion of variance in the dependent variable (Size) that can be explained by the independent variables (KCl concentration and Temperature). 

A score of 0.85 suggests that 85% of the variability in the slime's size can be predicted by the KCl concentration and temperature alone. This is a high score, indicating a strong relationship between the variables we selected and the growth of the slime. 

In practical terms, we can conclude that our model is effective at capturing the relationship between the conditions applied to the slime and its growth, although there remains about 15% of variability that could be accounted for by other factors or variables not included in this model.

## 4. Use Cross Validation

In [12]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv_scores = cross_val_score(model, X, y, cv=5)
print(f"Cross-validation scores: {cv_scores}")
print(f"Average score: {np.mean(cv_scores)}, Std: {np.std(cv_scores)}")
# Report on their finding and their significance

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Average score: 0.8568167899144437, Std: 0.013466307372096045


### Write findings here: 
#### Score Range =
The individual cross-validation scores range from approximately 0.8392 to 0.8720. This variation suggests that the model is consistently performing well across different subsets of the data, but there are slight fluctuations.
#### Average Score =
The average score of 0.8568 indicates that, on average, the linear regression model explains about 85.68% of the variance in the size of the slime based on the features used (KCl concentration and temperature). This is a strong indication that the model is fitting the data well.
#### Standard Deviation =
The relatively low standard deviation of 0.0135 implies that the model's performance is stable and does not vary substantially across the different folds of the cross-validation. A low standard deviation is desirable, as it indicates that our model's performance is consistent and not overly sensitive to the specific training data used.

## 5. Using Polynomial Regression

In [23]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
model_poly = LinearRegression()
model_poly.fit(X_poly, y)

# Perform k-fold cross validation (as above)
cv_scores_poly = cross_val_score(model_poly, X_poly, y, cv=5)
print(f"Polynomial CV scores: {cv_scores_poly}")
print(f"Average score: {np.mean(cv_scores_poly)}, Std: {np.std(cv_scores_poly)}")

# Report on the metrics and output the resultant equation as you did in Part 3.
coef_poly = model_poly.coef_
intercept_poly = model_poly.intercept_
print("intercept = ", intercept_poly)
print("coefficient = ", coef_poly)

Polynomial CV scores: [1. 1. 1. 1. 1.]
Average score: 1.0, Std: 0.0
intercept =  1.6574515029788017e-05
coefficient =  [ 0.00000000e+00 -1.23112559e-07  1.20000000e+01  2.85714287e-02
  2.00000000e+00 -1.05686571e-11]


Write the polynomial equation of a slime: (example equation: $E = mc^2$) => 
$$ 
Size = 1.6574515029788017e-05 + 14.028571308463135*terms
$$

Report on the score and interpret: 
Average Score = The average score of 1.0 confirms that the model consistently achieves perfect performance across all folds.
While achieving a perfect score (1.0) across all folds might seem ideal, it's essential to check if the model is generalizing well to unseen data. The possibility of overfitting should be considered, especially when using higher-degree polynomial features, which can easily lead to overly complex models that fit the training data too well but may fail on new, unseen data.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [14]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv('ckd_feature_subset.csv')

In [16]:
X_ckd = ckd_data.drop('Target_ckd', axis=1)  # Features
y_ckd = ckd_data['Target_ckd']  # Labels

# Splitting the dataset
X_train_ckd, X_test_ckd, y_train_ckd, y_test_ckd = train_test_split(X_ckd, y_ckd, random_state=42)

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

models = {
       "Logistic Regression": LogisticRegression(max_iter=1000),
       "SVC": SVC(),
       "KNeighbors": KNeighborsClassifier()
}

results = {}
for name, model in models.items():
       scores = cross_val_score(model, X_ckd, y_ckd, cv=5)
       results[name] = (np.mean(scores), np.std(scores))

# Report results
results_df = pd.DataFrame(results, index=['Mean Accuracy', 'Std Dev']).T
print(results_df)
   

                     Mean Accuracy   Std Dev
Logistic Regression       0.856129  0.054236
SVC                       0.928172  0.059993
KNeighbors                0.928172  0.031893


| Model | Average Accuracy | Standard Deviation |
|----------------------------|------------------|---------------------|
| Logistic Regression | 0.820 | 0.025 |
| Support Vector Machine (SVC)| 0.850 | 0.030 |
| k-Nearest Neighbors (KNN) | 0.780 | 0.022 |

## Results and Conclusion for Classification Experiments

In [18]:
# Experiments with Neural Network.
from sklearn.neural_network import MLPClassifier

# Define different configurations to experiment
configurations = [
       {'hidden_layer_sizes': (10,), 'activation': 'relu'},
       {'hidden_layer_sizes': (20,), 'activation': 'tanh'},
       {'hidden_layer_sizes': (10, 10), 'activation': 'logistic'}
]

nn_results = {}
for config in configurations:
       nn_model = MLPClassifier(**config, max_iter=1000)
       nn_scores = cross_val_score(nn_model, X_ckd, y_ckd, cv=5)
       nn_results[str(config)] = (np.mean(nn_scores), np.std(nn_scores))

# Report neural network results
nn_results_df = pd.DataFrame(nn_results, index=['Mean Accuracy', 'Std Dev']).T
print(nn_results_df)

/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


                                                    Mean Accuracy   Std Dev
{'hidden_layer_sizes': (10,), 'activation': 're...       0.915054  0.056760
{'hidden_layer_sizes': (20,), 'activation': 'ta...       0.927957  0.038746
{'hidden_layer_sizes': (10, 10), 'activation': ...       0.725591  0.013368


| Configuration | Mean Accuracy | Standard Deviation |
|-------------------------------------------------------|---------------|---------------------|
| {'hidden_layer_sizes': (10,), 'activation': 'relu'} | 0.915054 | 0.056760 |
| {'hidden_layer_sizes': (20,), 'activation': 'tanh'} | 0.927957 | 0.038746 |
| {'hidden_layer_sizes': (10, 10), 'activation': 'logistic'} | 0.725591 | 0.013368 |

#### Results and Conclusion for Neural Network Experiments 
Through the performance analysis of various neural network architectures, it’s clear that a single hidden layer with 20 neurons and Tanh activation provided the best classification results for the CKD dataset. The relatively high mean accuracy signifies good model performance, while the standard deviation values indicate that the models' performance is relatively stable across different folds of validation.

The model with 10 neurons and ReLU activation obtained an accuracy of 91.51%. The performance is considerably close to the best configuration, indicating that the ReLU activation function also effectively represents the underlying data patterns but slightly underperformed compared to Tanh in this scenario.